controller.go

```diff
@@ -1,95 +1,251 @@
 package funcmock
 
 import (
+	"fmt"
 	"reflect"
+	"sync"
 )
 
+type call struct {
+	params      []interface{}
+	returns     []interface{}
+	specReturns []reflect.Value
+}
+
+func (this *call) recordParams(vals []reflect.Value) {
+	this.params = make([]interface{}, len(vals))
+	for idx, val := range vals {
+		this.params[idx] = val.Interface()
+	}
+}
+
 type MockController struct {
-	originalFunc reflect.Value
-	targetFunc   reflect.Value
-	// we need map, not slice, to set call before it is called
-	callStack chan map[int]*call
-	// we need it to set call, before it is called
-	counter chan int
-	// the default call which shall be used for mint calls
-	defaultYield []reflect.Value
+	target, original reflect.Value
+	calls            []call
+	callCount        int
+	defaultReturns   []reflect.Value
+	lock             sync.Mutex
+	behaviour        *reflect.Value
+	preRecord        *reflect.Value
+	preReturn        *reflect.Value
+}
 
-	// Flag indicating the default return has been set
-	yieldSet bool
+type callHandle struct {
+	controller *MockController
+	calln      int
 }
 
-func (this *MockController) CallCount() int {
-	var count int
-	select {
-	case count = <-this.counter:
-		go func() { this.counter <- count }()
+func (this *MockController) getCall(calln int) *call {
+	if len(this.calls) <= calln {
+		//this growth factor guarantees amortized O(1) time for insertions
+		this.calls = append(this.calls, make([]call, calln+1)...)
```

Comment:
Since `getCall` is likely to be called with strictly increasing values in `calln`, `calls` is resized to at least twice its previous size. It makes the branch for resizing `calls` be taken at most log_2(n) times, reducing the times `append` and `make` are called.

Adding clarifying comment.